src/Components/Web/src/Routing/NavigationLock.cs

```diff
@@ -38,26 +38,45 @@ public sealed class NavigationLock : IComponent, IAsyncDisposable
 
     void IComponent.Attach(RenderHandle renderHandle)
     {
+        _renderHandle = renderHandle;
     }
 
-    async Task IComponent.SetParametersAsync(ParameterView parameters)
+    Task IComponent.SetParametersAsync(ParameterView parameters)
     {
-        var lastHasOnBeforeInternalNavigationCallback = HasOnBeforeInternalNavigationCallback;
-        var lastConfirmExternalNavigation = ConfirmExternalNavigation;
+        foreach (var parameter in parameters)
+        {
+            if (parameter.Name.Equals(nameof(OnBeforeInternalNavigation), StringComparison.OrdinalIgnoreCase))
+            {
+                OnBeforeInternalNavigation = (EventCallback<LocationChangingContext>)parameter.Value;
+            }
+            else if (parameter.Name.Equals(nameof(ConfirmExternalNavigation), StringComparison.OrdinalIgnoreCase))
+            {
+                ConfirmExternalNavigation = (bool)parameter.Value;
+            }
+            else
+            {
+                throw new ArgumentException($"The component '{nameof(NavigationLock)}' does not accept a parameter with the name '{parameter.Name}'.");
+            }
+        }
 
-        parameters.SetParameterProperties(this);
+        _renderHandle.Render(static builder => { });
+        return Task.CompletedTask;
+    }
 
-        var hasOnBeforeInternalNavigationCallback = HasOnBeforeInternalNavigationCallback;
-        if (hasOnBeforeInternalNavigationCallback != lastHasOnBeforeInternalNavigationCallback)
+    async Task IHandleAfterRender.OnAfterRenderAsync()
+    {
+        var lastHasLocationChangingHandler = _locationChangingRegistration is not null;
+        var hasLocationChangingHandler = OnBeforeInternalNavigation.HasDelegate;
+        if (lastHasLocationChangingHandler != hasLocationChangingHandler)
         {
             _locationChangingRegistration?.Dispose();
-            _locationChangingRegistration = hasOnBeforeInternalNavigationCallback
+            _locationChangingRegistration = hasLocationChangingHandler
                 ? NavigationManager.RegisterLocationChangingHandler(OnLocationChanging)
                 : null;
         }
```

Comment:
It's not necessarily a problem, but it seems a bit unusual to trigger rendering unconditionally, and then be doing all this state management during `OnAfterRenderAsync`.

Would it work to update these fields during `SetParametersAsync` and then only trigger a render (and hence the after-render callback) during the rare case when we need to make the JS interop call because the flag has changed?

BTW this is not a big deal at all, so if you feel you have more urgent things to focus on, I'm happy to sign off on the implementation as it stands here. Totally up to you whether you want to tweak this further.

Reply:
This makes sense. I just changed the logic to only re-render when the parameter values change.

However, I decided to keep some of the state management logic in `OnAfterRenderAsync()` in order to minimize the separation between updating the fields tracking the "last applied" parameter values and actually applying the parameter values (e.g. making JS interop calls). I figured that would minimize the chance for some incorrect code path to be taken in between updating fields from `SetParametersAsync()` and applying them in `OnAfterRenderAsync()`.